### PR TITLE
Add support for wallpaper type Picture of the Day

### PR DESCRIPTION
**Pull request by abhr-0**

### CODE REVIEW

File: modules/workspace.nix [hunk 1]
```diff
@@ -99,14 +108,27 @@ in
         Allows you to set wallpaper slideshow. Needs a directory of your wallpapers and an interval length.
       '';
     };
+
+    wallpaperPictureOfTheDay = lib.mkOption {
+      type = lib.types.nullOr wallpaperPictureOfTheDayType;
+      default = null;
+      example = "apod";
+      description = ''
+        Allows you to set wallpaper using the picture of the day plugin. Needs the provider and the fill mode.
```
Comment (magnouvean) [inline modules/workspace.nix]: I don't see anything to do with the fill-mode in this pr. Is something missing?

**Review by magnouvean — COMMENTED** [review]
Looks good otherwise. Will test soon.

**Comment by magnouvean** [comment]
Just tested and it does work well! Could just set the type to `nullOr (enum ["apod" "bing" "flickr" "natgeo" "noaa" "wcpotd" "epod" "simonstalenhag" ])` I reckon. Works fine as is now without fill-mode functionality.

File: modules/workspace.nix [hunk 2]
```diff
@@ -99,14 +108,27 @@ in
...
+
   };
 
   config = (lib.mkIf cfg.enable (lib.mkMerge [
     {
       assertions = [
         {
-          assertion = (cfg.workspace.wallpaperSlideShow == null || cfg.workspace.wallpaper == null);
-          message = "Cannot set both wallpaper and wallpaperSlideShow at the same time.";
+          assertion = ((cfg.workspace.wallpaperSlideShow == null && cfg.workspace.wallpaper == null && cfg.workspace.wallpaperPictureOfTheDay != null)
```
Comment (pluiedev) [inline modules/workspace.nix]: Consider doing something like this instead: (see https://github.com/nix-community/home-manager/blob/master/modules/programs/git.nix#L360-L365)
```nix
assertion = let 
  wallpapers = with cfg.workspace; [wallPaperSlideShow wallpaper wallpaperPictureOfTheDay];
in
  lib.count (x: x != null) wallpapers <= 1;
```

Additionally, this assertion currently triggers when none of the three are set — are we really going to force people to pick a wallpaper?

File: modules/workspace.nix [hunk 3]
```diff
@@ -99,14 +108,27 @@ in
         Allows you to set wallpaper slideshow. Needs a directory of your wallpapers and an interval length.
       '';
     };
+
+    wallpaperPictureOfTheDay = lib.mkOption {
+      type = lib.types.nullOr wallpaperPictureOfTheDayType;
+      default = null;
+      example = "apod";
```
Comment (pluiedev) [inline modules/workspace.nix]: Wait... I don't think you can assign strings to submodules. Honestly, the submodule isn't even necessary and a simple enum type would do just fine

File: modules/workspace.nix [hunk 4]
```diff
@@ -170,5 +192,19 @@ in
         priority = 3;
       };
     })
+    (lib.mkIf (cfg.workspace.wallpaperPictureOfTheDay != null) {
+      programs.plasma.startup.desktopScript."set_wallpaper_potd" = {
+        text = ''
+          let allDesktops = desktops();
+          for (var desktopIndex = 0; desktopIndex < allDesktops.length; desktopIndex++) {
+              var desktop = allDesktops[desktopIndex];
```
Comment (pluiedev) [inline modules/workspace.nix]: It's currently the year 2024. Plasma's desktop scripts support ES6, even though all their docs are ES5 for some reason
```js
for (const desktop of desktops()) {
    // ...
}
```

**Review by pluiedev — CHANGES_REQUESTED** [review]

File: modules/workspace.nix (hunk 1 above)
Comment (abhr-0) [reply modules/workspace.nix]: Oh sorry, actually I had added support for fill-mode but specifically just for potd, but I realised that we should be able to set it for all types of wallpapers. So I removed that bit (cuz that should be a different PR) but ig I forgot to update the description. Whoops.

File: modules/workspace.nix (hunk 2 above)
Comment (abhr-0) [reply modules/workspace.nix]: > ```nix
> assertion = let 
>   wallpapers = with cfg.workspace; [wallPaperSlideShow wallpaper wallpaperPictureOfTheDay];
> in
>   lib.count (x: x != null) wallpapers <= 1;
> ```

This! I really hated the way I did it, but I just couldn't figure out how to make it nicer.

> Additionally, this assertion currently triggers when none of the three are set — are we really going to force people to pick a wallpaper?

Oops. that's definitely wasn't what I wanted to do.

Big thanks for the help

File: modules/workspace.nix (hunk 3 above)
Comment (abhr-0) [reply modules/workspace.nix]: Ah, yes that's completely true, I'll update the example.

**Comment by abhr-0** [comment]
Big thanks to @magnouvean and @pluiedev for the help.

Honestly, I don't have much experience with nix (or even git) . I apologize for all the goof ups.

File: modules/workspace.nix [hunk 5]
```diff
@@ -99,14 +108,27 @@ in
         Allows you to set wallpaper slideshow. Needs a directory of your wallpapers and an interval length.
       '';
     };
+
+    wallpaperPictureOfTheDay = lib.mkOption {
+      type = lib.types.nullOr wallpaperPictureOfTheDayType;
+      default = null;
+      example = ''{ provider = "apod" }'';
```
Comment (magnouvean) [inline modules/workspace.nix]: Should be `{ provider = "apod"; }` instead (i.e. remove the quotes and add semicolon after provider).

**Review by magnouvean — COMMENTED** [review]
No worries, it's a good way to learn (and this helps get some new features for the project that I think I might even start to use myself).

Anyways this looks good to me now once the adjustment in this comment is fixed. Is it a possibility that other options (other than provider) is added to `wallpaperPictureOfTheDay`? If not we could just set the type of `wallpaperPictureOfTheDay` to `nullOr (enum ["apod" "bing" "flickr" "natgeo" "noaa" "wcpotd" "epod" "simonstalenhag" ]);` directly, and delete `wallpaperPictureOfTheDayType` altogether. I'm not opposed to having this here though if it makes sense to extend one day, as the current type does make sense.

File: modules/workspace.nix (hunk 5 above)
Comment (abhr-0) [reply modules/workspace.nix]: Well, POTD seems to have another option to enable/disable updating over a metered connection. Adding it immediately.

> Should be `{ provider = "apod"; }` instead (i.e. remove the quotes and add semicolon after provider).

Even after being told that the example was not supposed to be a string, I fixed it...by changing it to another string. :sweat_smile:

**Comment by magnouvean** [comment]
Very nice! I will test and hopefully merge soon if everything works fine

File: modules/workspace.nix [hunk 6]
```diff
@@ -170,5 +197,20 @@ in
         priority = 3;
       };
     })
+    (lib.mkIf (cfg.workspace.wallpaperPictureOfTheDay != null) {
+      programs.plasma.startup.desktopScript."set_wallpaper_potd" = {
+        text = ''
+          let allDesktops = desktops();
+          for (const desktop of allDesktops) {
+              var desktop = allDesktops[desktopIndex];
```
Comment (pluiedev) [inline modules/workspace.nix]: This doesn't compile as you redefine desktop, and also, desktopIndex is undefined

File: modules/workspace.nix [hunk 7]
```diff
@@ -170,5 +197,20 @@ in
...
+              desktop.wallpaperPlugin = "org.kde.potd";
+              desktop.currentConfigGroup = Array("Wallpaper", "org.kde.potd", "General");
```
Comment (pluiedev) [inline modules/workspace.nix]: Use array literal syntax
```js
["Wallpaper", "org.kde.potd", "General"]
```
I know that KDE docs use the Array(...) form but it's really just outdated. Writing modern JS works just fine.

**Review by pluiedev — CHANGES_REQUESTED** [review]

**Review by pluiedev — APPROVED** [review]

**Comment by magnouvean** [comment]
Just tested and works as expected still. Good work. LGTM

**Pull request merged**